### PR TITLE
Add isLoading to useProductMetadata hook

### PR DESCRIPTION
**Pull request by nathanss**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This adds a `isLoading` property to the useProductMetadata hook, in order for consumers to know if the data has finished loading or not.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Pretty simple change. In my opinion, no manual tests are needed. I tested it with developing an extension and also added an unit test.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

**Review by retrofox — COMMENTED** [review]
I'd like to suggest moving the logic that isn't about selecting data (the `updateMetadata` helper) out of the select body function.

Also, `updateMetadata` sounds a little bit redundant to me since, in the end, the hook is about the metadata. Is it too late to rename it to `update`:

```jsx

function MyComponent() {
  const { metadata, update } = useProductMetadata();
 
  return (
    <Button onClick={ () => update( { name: 'Finn' } ) }>
      { __( 'Update name', 'woocommerce' ) }
    </Button>
  );
}
```
Please feel free to ignore the comment.

**Comment by nathanss** [comment]
@retrofox I prefer keeping the update function inside the `useSelect`, since it uses the metadata array, which is not exposed. I've renamed some variables to make this clearer.

I agree on the `updateMetadata` suggestion. I've renamed it!

File: packages/js/product-editor/src/hooks/use-product-metadata.ts [hunk 1]
```diff
@@ -25,38 +25,40 @@ function useProductMetadata( options?: Options ) {
...
+								entries.findIndex(
+									( e ) => e.key === item.key
+								) === -1
+						),
+						...entries,
+					],
+				} ),
+			isLoading: ! isResolutionFinished,
+		};
+	} );
```
Comment (retrofox) [inline packages/js/product-editor/src/hooks/use-product-metadata.ts]: I see this select instance doesn't have any defined dependency. Is it ok?

**Comment by retrofox** [comment]
> @retrofox I prefer keeping the update function inside the `useSelect`, since it uses the metadata array, which is not exposed. I've renamed some variables to make this clearer.

The `metadata` array is available in the hook function body. You can pick them and use them in the update handler.

Also, and like a general rule, the select should take over only selecting and returning data.
For instance, in this case, select should return the metadata array. Then, out of the select function body, process the array to create the object. Probably in this implementation, this _Array-to-Object_ process could generate unexpected renders in the component.

File: packages/js/product-editor/src/hooks/use-product-metadata.ts (hunk 1 above)
Comment (retrofox) [reply packages/js/product-editor/src/hooks/use-product-metadata.ts]: I see you removed the prev `[ id ]` dependencies array from the second argument of the select

**Comment by retrofox** [comment]
> **Be careful about transforming data inside the callback**
> The following useSelect call will have suprising behavior. It will cause the component to rerender on each update in the taxonomies store, even if the tags haven’t changed at all:
>
>
> ```es6
> const { tagNames } = useSelect( ( select ) => {
>  const tags = select( 'taxonomies' ).getTags();
>  return { tagNames: tags.map( ( t ) => t.name ) };
>}, [] );
>```
>
> This happens because every invocation of the callback returns a different array (the result of .map) even though the raw tags array in the store hasn’t changed. Because the returned array is not equal (===) to the previous value, it is detected as a change that needs to trigger a re-render.
>
> The solution to this problem is to move the data transformation outside the useSelect hook, and wrap it in useMemo:
>
>```es6
> const { tags } = useSelect( ( select ) => ( {
>  tags: select( 'taxonomies' ).getTags(),
>} ), [] );
>
> const tagNames = useMemo( () => {
>  return tags.map( ( t ) => t.name );
> }, [ tags ] );
>```
> This will cause a re-render only when the raw tags really change.

Part of [Best practices for using useSelect() from @wordpress/data](https://jardasn.blog/2021/01/22/some-best-practices-for-using-useselect-from-wordpress-data/)

Is it a similar scenario for the `metadata` object?

**Review by retrofox — APPROVED** [review]
LGTM. Thanks for handling the feedback 🙇‍♂️ 
Probably, it would be nice to memorize the `meta_data` array but that's something we can do in a follow-up. 🚀

**Pull request merged**